ruby/lib/ci/queue/redis/supervisor.rb

```diff
@@ -22,16 +22,28 @@ def wait_for_workers
           yield if block_given?
 
           time_left = config.timeout
-          until exhausted? || time_left <= 0 || max_test_failed?
-            sleep 1
+          time_left_with_no_workers = config.queue_init_timeout
+          until exhausted? || time_left <= 0 || max_test_failed? || time_left_with_no_workers <= 0
             time_left -= 1
 
+            if active_workers?
+              time_left_with_no_workers = config.queue_init_timeout
+            else
+              time_left_with_no_workers -= 1
+            end
+
             yield if block_given?
           end
           exhausted?
         rescue CI::Queue::Redis::LostMaster
           false
         end
+
+        private
+
+        def active_workers?
+          redis.zrangebyscore(key('running'), Time.now.to_f - config.timeout, "+inf").count > 0
```

Comment:
Minor micro-opt, you can pass `limit: [0, 1]` to get returned one record at most.